### PR TITLE
Begin PETSc recipes

### PR DESCRIPTION
**Pull request by minrk**
sourced from https://bitbucket.org/petsc/conda-recipes/

changes from original:

- remove MPI feature for now (only use mpich, as it's the only MPI packaged by conda-forge pending #557)
- reorder sections according to recipe lint

Perhaps @dalcinl would be interested in moving those recipes over here.

### CODE REVIEW

**Comment by ocefpaf** [comment]
@minrk I am working on a [recipe](https://github.com/ocefpaf/staged-recipes/tree/openmpi) for `openmpi` based on `petsc` one you referenced in #557. Should have something for you soon.

**Comment by minrk** [comment]
@ocefpaf cool, thanks!

**Comment by minrk** [comment]
Hm, malformed things going on on OS X, will investigate.

File: recipes/petsc/build.sh [hunk 1]
```diff
@@ -0,0 +1,47 @@
+#!/bin/bash
+
+export PETSC_DIR=$SRC_DIR
+export PETSC_ARCH=arch-conda-c-opt
+
+python2 ./configure \
```
Comment (patricksnape) [inline recipes/petsc/build.sh]: Should this be `$PYTHON`?

Comment (minrk) [inline recipes/petsc/build.sh]: Yes, indeed. Updated to match.

File: recipes/petsc/meta.yaml [hunk 2]
```diff
@@ -21,7 +21,7 @@ build:
 
 requirements:
   build:
-    - python
+    - python <3
```
Comment (patricksnape) [inline recipes/petsc/meta.yaml]: Generally we enforce this with the skip preprocessor - so above where you skip windows we would also skip python<3 -> `skip: true  # [win or py2k]`

Comment (minrk) [inline recipes/petsc/meta.yaml]: gotcha, done that way, now.

**Comment by minrk** [comment]
I'm not sure what's going on with the Travis failure. When I build on my local macs, it's working fine. I'll have to come back to it, unless someone has a good idea.

File: recipes/petsc/build.sh [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+make
+make install
+
+if [[ $(uname) == Darwin ]];
+then
+    library=$PREFIX/lib/lib$PKG_NAME.dylib
+    pathlist=$(otool -l $library | grep ' path /' | awk '{print $2}')
+    for path in $pathlist; do
+        install_name_tool -delete_rpath $path $library
```
Comment (patricksnape) [inline recipes/petsc/build.sh]: Why is this happening?

**Comment by dfroger** [comment]
Hello, I don't know if that can be usefull, [here](https://github.com/sed-pro-inria/conda-recipes/tree/master/petsc) is a conda recipe for `PETSc` that I wanted to integrate in `conda-forge` (when I have the time)...

**Comment by dalcinl** [comment]
@minrk Sure, I'm really interested in moving everything here. However, we have to get the MPI thing ready before. That means we all should agree that some people prefer MPICH, others prefer Open MPI, then we have to use conda features and provide builds of MPI-based packages built with each implementation.

File: recipes/petsc/build.sh (hunk 3 above)
Comment (minrk) [inline recipes/petsc/build.sh]: I'm not sure, and it doesn't fail on my (10.11) system. @dalcinl do you know what this rpath-deleting solves? Perhaps recent conda-builds make this obsolete?

**Comment by minrk** [comment]
@dalcinl great! mpich is packaged already, and @ocefpaf is working on openmpi in #1490. Once that's in, I was planning on following the example of adding an mpi metapackage like is already used [for blas](https://github.com/conda-forge/blas-feedstock/blob/master/recipe/meta.yaml), and following that example of numpy, etc.

BTW, @dalcinl shall I add you as a maintainer on these packages?

**Comment by minrk** [comment]
I skipped the delete_rpath step and the package successfully builds, but I cannot speak to what that might break.

**Comment by patricksnape** [comment]
conda-build already fixes rpath so I don't think that was even required.

File: recipes/petsc/build.sh [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+make
+make install
+
+if [[ $(uname) == Darwin ]];
+then
+    library=$PREFIX/lib/lib$PKG_NAME.dylib
+else
+    library=$PREFIX/lib/lib$PKG_NAME.so
+fi
+$PYTHON $RECIPE_DIR/replace-binary.py $(dirname $SRC_DIR) "" $library
```
Comment (patricksnape) [inline recipes/petsc/build.sh]: I'm not sure this is needed either because conda-build also fixes binary prefixes.

**Comment by minrk** [comment]
Okay, this now removes the recipe's replacements, trusting conda-build to do the right thing, and links openblas instead of mkl.

File: recipes/petsc/build.sh [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+sedinplace() { [[ $(uname) == Darwin ]] && sed -i "" $@ || sed -i"" $@; }
+for path in $PETSC_DIR $PREFIX; do
+    sedinplace s%$path%\${PETSC_DIR}%g $PETSC_ARCH/include/petsc*.h
+done
+
+make
+make install
+
+rm -fr $PREFIX/bin
```
Comment (patricksnape) [inline recipes/petsc/build.sh]: Do we really want to delete the entire `bin` and `share` directories? Or are there just particular files that PETSc introduces?

**Comment by patricksnape** [comment]
OK this is starting to look good to me - this is quite a complex recipe and so would appreciate a second pair of eyes from someone: cc @conda-forge/core

**Comment by dalcinl** [comment]
@minrk Yes, please, add me as an admin of these packages.

File: recipes/petsc/build.sh (hunk 5 above)
Comment (dalcinl) [inline recipes/petsc/build.sh]: IMHO, the ``bin`` and ``share`` directories are a bit messy. My intention at the time I wrote these recipes was to build a minimal package with just the stuff required to build dependencies like (``petsc4py``, ``SLEPc`` and ``slepc4py``)

File: recipes/petsc/build.sh [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  --with-ssl=0 \
+  --with-x=0 \
+  --prefix=$PREFIX
+
+sedinplace() { [[ $(uname) == Darwin ]] && sed -i "" $@ || sed -i"" $@; }
+for path in $PETSC_DIR $PREFIX; do
+    sedinplace s%$path%\${PETSC_DIR}%g $PETSC_ARCH/include/petsc*.h
+done
+
+make
```
Comment (jakirkham) [inline recipes/petsc/build.sh]: Is there a `make check` we can run?

File: recipes/petsc/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    - blas 1.* {{blas}}
+    - {{mpi}} {{req[mpi]}}
+  run:
+    - blas 1.* {{blas}}
+    - {{mpi}} {{req[mpi]}}
+
+about:
+  home: http://www.mcs.anl.gov/petsc/
+  summary: 'PETSc: Portable, Extensible Toolkit for Scientific Computation'
+  license: 'BSD'
```
Comment (jakirkham) [inline recipes/petsc/meta.yaml]: What version of BSD (e.g. `BSD 3-Clause`)? Also is there a [license file]( http://conda.pydata.org/docs/building/meta-yaml.html#license-file ) that we can include?

File: recipes/petsc/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+extra:
+  recipe-maintainers:
+    # Always check with the people listed below if they are OK becoming maintainers of the recipe. (There will be spam!)
```
Comment (jakirkham) [inline recipes/petsc/meta.yaml]: nit: Please drop this example comment.

File: recipes/petsc4py/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    - petsc4py.PETSc
+
+about:
+  home: https://bitbucket.org/petsc/petsc4py
+  summary: Python bindings for PETSc
+  license: BSD
+
+extra:
+  recipe-maintainers:
+    # Always check with the people listed below if they are OK becoming maintainers of the recipe. (There will be spam!)
```
Comment (jakirkham) [inline recipes/petsc4py/meta.yaml]: nit: Please drop this example comment.

File: recipes/petsc/meta.yaml [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+build:
+  number: {{build}}
+  binary_relocation: true
+  detect_binary_files_with_prefix: true
+  skip: true  # [win or py3k]
+  features:
+    - blas_{{blas}}
+
+requirements:
+  build:
```
Comment (jakirkham) [inline recipes/petsc/meta.yaml]: Please add `toolchain` to the build requirements.

File: recipes/petsc4py/meta.yaml [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+  md5: {{md5}}
+
+build:
+  number: {{build}}
+  script:
+    - '$PYTHON setup.py build --petsc-dir=$PREFIX install'
+  skip: true  # [win]
+
+requirements:
+  build:
```
Comment (jakirkham) [inline recipes/petsc4py/meta.yaml]: Please add `toolchain` to the build requirements.

File: recipes/petsc4py/meta.yaml [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    - python
+    - numpy
```
Comment (dalcinl) [inline recipes/petsc4py/meta.yaml]: NumPy is a compile-time  and runtime dependency of this package.

File: recipes/petsc4py/meta.yaml (hunk 11 above)
Comment (minrk) [inline recipes/petsc4py/meta.yaml]: Done. For reference for my other recipes, what does toolchain do and under what circumstances should I use it? All recipes that compile anything?

**Comment by minrk** [comment]
I think this is good to go, unless there's more discussion required about the cleanup of bin/share.

File: recipes/petsc/build.sh (hunk 5 above)
Comment (ocefpaf) [inline recipes/petsc/build.sh]: I would keep them but ultimately that is up to the maintainers of this recipe.

**Pull request merged**